src/recordlinker/splunk.py

```diff
@@ -0,0 +1,79 @@
+import json
+import time
+import typing
+import urllib.parse
+import urllib.request
+
+TIMEOUT = 5
+
+
+class SplunkError(Exception):
+    pass
+
+
+class SplunkHECClient:
+    PATH = "/services/collector/event"
+
+    def __init__(self, splunk_uri: str) -> None:
+        """
+        Create a new Splunk HEC client and test its connection.
+        The URI uses a custom scheme to specify the Splunk HEC server and parameters.
+        The URI format is:
+            splunkhec://<token>@<host>:<port>?index=<index>&proto=<protocol>&ssl_verify=<verify>&source=<source>
+        """
+        try:
+            uri: urllib.parse.ParseResult = urllib.parse.urlparse(splunk_uri)
+            # flatten the query string values from lists to single values
+            qs: dict[str, str] = {k: v[0] for k, v in urllib.parse.parse_qs(uri.query).items()}
+
+            if uri.scheme != "splunkhec":
+                raise SplunkError(f"invalid scheme: {uri.scheme}")
+
+            scheme = qs.get("proto", "https").lower()
+            host = f"{uri.hostname}:{uri.port}" if uri.port else uri.hostname
+            self.url = f"{scheme}://{host}{self.PATH}"
+            self.headers = {
+                "Authorization": f"Splunk {uri.username}",
+                "Content-Type": "application/json",
+            }
+            # initialize the default payload parameters
+            self.params: dict[str, str] = {"host": uri.hostname or "", "sourcetype": "_json"}
+            if qs.get("index"):
+                self.params["index"] = qs["index"]
+            if qs.get("source"):
+                self.params["source"] = qs["source"]
+            self._test_connection()
+        except Exception as exc:
+            raise SplunkError(f"invalid connection: {splunk_uri}") from exc
+
+    def _send_request(self, body: bytes | None = None):
+        request = urllib.request.Request(self.url, data=body, method="POST", headers=self.headers)
+        try:
+            with urllib.request.urlopen(request, timeout=TIMEOUT) as response:
+                # return the response status code
+                return response.getcode()
+        except urllib.error.HTTPError as exc:
+            return exc.code
+
+    def _test_connection(self) -> None:
+        status = self._send_request()
+        # check for a 400 bad request, which indicates a successful connection
+        # 400 is expected because the payload is empty
+        if status != 400:
```

Comment:
Is there a splunk health check endpoint that could be checked to test the connection?

Reply:
Possibly.  I'll have to use a different endpoint, but I'll test this out https://community.splunk.com/t5/Getting-Data-In/How-to-check-if-an-HEC-is-up-or-not-before-posting-any-data-to/m-p/417404

Reply:
So a call like `curl -X POST http://localhost:8088/services/collector -H "Authorization: Splunk $HECTOKEN" -d ''`, works in that the response comes back as `{"text":"No data","code":5}` but the response code is still 400.  I could check the response payload rather than the status code, but the "health check" still results in a 400 bad request either way.

Reply:
Okay yeah wasn't sure if there was already an endpoint for this type of check. Makes sense to leave it as you have it then.